core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -166,9 +166,11 @@ private[spark] class MetricsSystem private (
   }
 
   def removeSource(source: Source) {
-    sources -= source
-    val regName = buildRegistryName(source)
-    registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    if (sources.contains(source)) {
+      sources -= source
```

Comment:
Is it really necessary to navigate `sources` twice - the first during `contains`, the 2nd - while removing it?

Reply:
I add existing check because original code removes the metrics no matter whether the source was registered or not. i will use index to avoid the twice navigating.thanks for your advice.